mlfinlab/data_structures/base_bars.py

```diff
@@ -16,74 +16,115 @@ class BaseBars(ABC):
     they are included here so as to avoid a complicated nested class structure.
     """
 
-    def __init__(self, file_path, metric, batch_size=2e7):
+    def __init__(self, metric, batch_size=2e7, verbose=True, to_csv=False, output_path=None, cache=None):
         """
         Constructor
 
-        :param file_path: (String) Path to the csv file containing raw tick data in the format[date_time, price, volume]
-        :param metric: (String) type of imbalance bar to create. Example: dollar_imbalance.
-        :param batch_size: (Int) Number of rows to read in from the csv, per batch.
+        :param metric: (String) type of imbalance bar to create. Example: dollar_imbalance
+        :param batch_size: (Int) Number of rows to read in from the csv, per batch
+        :param verbose: (Boolean) Flag whether to print message on each processed batch or not
+        :param to_csv: (Boolean) Flag for writing the results of bars generation to local csv file, or to in-memory DataFrame
+        :param output_path: (String) Path to results file, if to_csv = True
+        :param cache: (List) Path to results file, if to_csv = True
         """
         # Base properties
-        self.file_path = file_path
         self.metric = metric
         self.batch_size = batch_size
+        self.verbose = verbose
+
         self.prev_tick_rule = 0
 
-        # Batch_run properties
+        # Run properties
+        self.cols = ['date_time', 'open', 'high', 'low', 'close', 'volume']
+
+        self.to_csv = to_csv
+        self.output_path = output_path
+
+        if self.to_csv is True:
+            open(self.output_path, 'w').close()  # clean output csv file
+
+        self.count = 0
+        self.cache = cache or []
+
         self.flag = False  # The first flag is false since the first batch doesn't use the cache
-        self.cache = []
 
-    def batch_run(self, verbose=True, to_csv=False, output_path=None):
+    def _batch_iterator(self, data):
         """
-        Reads a csv file in batches and then constructs the financial data structure in the form of a DataFrame.
-        The csv file must have only 3 columns: date_time, price, & volume.
-        :param verbose: (Boolean) Flag whether to print message on each processed batch or not
-        :param to_csv: (Boolean) Flag for writing the results of bars generation to local csv file, or to in-memory DataFrame
-        :param output_path: (Boolean) Path to results file, if to_csv = True
+        Iterate depending on data type.
 
-        :return: (DataFrame or None) Financial data structure
+        :param data: (String or DataFrame or List) the dataset to be iterated
```

Comment:
As we have merged PR with updated data structures, this PR can deprecated.

Reply:
Do you have no intention to support input of type list or tuple?

Reply:
As for me, list and tuples can be easily converted to pd.DataFrame. However, list of csv files can be useful, but in this case PR needs to be updated to resolve conflicts with develop branch implementation

Reply:
I would be pleased to update this PR with develop branch. Lists or tuples are desirable when iterating on a live stream. However, type checking a list as either financial data input or a list of csv files could be problematic. I am using OrderedDict for live stream. How about list or tuple for CSV file input, and OrderedDict as an alternative to DataFrame?

Reply:
We would like not to replace DataFrame because it is widespread and used by many users

Reply:
Agreed. I do not wish to replace DataFrame, just to support additional inputs. On a live stream, instantiating a DataFrame for a single row is a bit heavy. So, you prefer to only support CSV files, and/or DataFrames?

Reply:
I think that would be great to support a structure close to live stream. So live stream, csv file (done), DataFrame(done), list of csv files will be perfect

Reply:
I updated the PR. I think Travis CI, fails on test_feature_importance and test_orthogonal_features. 

Some things should probably be renamed. I just wonder if you are satisfied with the general direction of the PR.

Reply:
The general direction seems good to me. One very important thing is not to validate values checks in unit tests to be sure that new feature doesn't change base algorithm performance

Reply:
As discussed, I had hoped to use this in a server(less) process. However, `import mlfinlab` also imports sklearn, numba, etc. so actually it's a non-starter. 

Avoiding the imports could be accomplished with namespace packages; however, I doubt your team is interested in maintaining that https://packaging.python.org/guides/packaging-namespace-packages/

So, I decided to use BigQuery with a FireStore cache, so wrote my own classes.

The small changes I made in this pull request do allow 1) live stream and 2) list of CSV files, as discussed. So, I suppose it complete.

What would you like to see with this pull request?

Reply:
Hi @globophobe, 

This looks rather interesting. We will review the code, it would, of course, be exciting to add a live data feed implementation. 

I see that all the unit tests pass and the pylinter passes so that's a good sign. 

Did you remember to update the documentation to reflect your changes? If we update the branch but not the docs then users will be confused by the new functionality. 

So far, excited by this.

Reply:
Revised docs: 

> It is for this reason that the mlfinlab package __requires__ a file path... 

To: 

> It is for this reason that the mlfinlab package __suggests using__ a file path...